### ISSUE TITLE
Obfuscated code behaves differently

### ISSUE THREAD

**Issue by lextudio-support** [obfuscar/obfuscar#505]
__EDIT:__ added a second piece of code that doesn't use explicitly implemented interface members

Consider the following code:

```
  class Program
    {
        [Obfuscation(Exclude = true)]
        static void Main(string[] args)
        {
            var t = new Two();
            Console.WriteLine(t.Property1);
            Console.WriteLine(t.Property2);
            Console.WriteLine(t.Property3);
            Console.WriteLine(((I1) t).Property1);
            Console.WriteLine(((I3) t).Property1);
            Console.WriteLine(((I3) t).Property3);
            Console.WriteLine(((I3) t).Property2);

            var o = new One();
            Console.WriteLine(o.Property1);
            Console.WriteLine(((I1) o).Property1);
        }
    }

    interface I1 { double Property1 { get; } }
    interface I2 { double Property2 { get; } }
    interface I3 : I1, I2 { double Property3 { get; } }

    class One : I1
    {
        public double Property1 { get { return 1; } }
        public double Another1 { get { return -1; } }
        double I1.Property1 { get { return 1.5; } }
    }

    class Two : One, I3
    {
        public double Property2 { get { return 2; } }
        public double Property3 { get { return 3; } }
        public double Another2 { get { return -2; } }
        public double Another3 { get { return -3; } }
        double I3.Property3 { get { return 3.5; } }
    }

```

Executed before obfuscation the output is:

```
1 2 3 1.5 1.5 3.5 2 1 1.5
```

Obfuscated with just KeepPublicApi=false and HidePrivateApi=true using the latest code in the repo (79ee199979), and executing the obfuscated assembly, the output changes:

```
1 2 3 **2** **2** 3.5 2 1 1.5
```

I've been fighting with the sourcecode trying to understand and fix (hack) this but I still haven't been lucky...

__EDIT__: this piece of code is simpler and also fails:

```
    class Program
    {
        [Obfuscation(Exclude = true)]
        static void Main(string[] args)
        {
            var x = new ClassB();
            Console.Write(x.PropB); // B
            Console.Write(x.PropA); // A
            Console.Write(((IB) x).PropB); // B
            Console.Write(((IB) x).PropA); // A
            Console.Write(((IA) x).PropA); // A
            Console.Write(((ClassA) x).PropA); // A
            Console.WriteLine();
        }
    }

    public interface IA { string PropA { get; } }
    public interface IB : IA { string PropB { get; } }

    public class ClassA : IA { public string PropA { get { return "A"; } } }
    public class ClassB : ClassA, IB { public string PropB { get { return "B"; } } }
```

Non-obfuscated output: BABAAA
Obfuscated output: BAB**BB**A

**Original Reported Date**: 2014-04-10T12:27:10.917-07:00
**Planned For Release**: 
**Original CodePlex ID**: 7

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

Will take a look at it in the next few days. Should be a priority issue which can be solved soon.

**Original Posted Date**: 2014-04-11T01:26:31.413-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

Thanks! Just added a much simpler example where there is no explicitly implemented interface members and it fails too.

**Original Posted Date**: 2014-04-11T01:31:54.18-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

The obfuscation process reuses names, so that it really confuses CLR on which method should be called at runtime. I am still investigating to see whether there is any workaround. Will post back if I have any finding.

**Original Posted Date**: 2014-04-29T05:37:26.623-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

1. There is BUG in renaming properties. Property is renamed, but not registred in method names. For example: a/ interface xxi {  int code {get;} }  b/ interface is not obfuscated c/you have class ZZ :xxi {..}, d/ method get_code is obfuscated in class ZZ. It cause runtime error.
Correct:
RenameMethods ()
...
  if (m.Status == ObfuscationStatus.Skipped) {
	// IMPORTANT: shortcut for event and property methods.
        // JIRKA:To avoid rename public property
        MethodGroup mgr = project.InheritMap.GetMethodGroup(methodKey);
        if (mgr == null || mgr.Name == null) RenameVirtualMethod(info, baseSigNames, sigNames, methodKey, method, m.StatusText);
	continue;
   }
But there is still error in result BABBBA. 
2. I changed code for rename property to hash property instead of generate it and it is ok.

**Original Posted Date**: 2014-05-22T19:35:20.173-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

Property are renamed to same name for example:
public interface IA { string PropA { get; } }
public interface IB : IA { string PropB { get; } }
TO:
public interface IA { string A { get; } }
public interface IB : IA { string A { get; } }
Both property have same name.
Quick corretion is to make counter for renaming global insted of for every class.
In RenameProperites method()
..
You have to replace this varibale to global for all "uniqueMemberNameIndex++"  i.e. instance varibale for class Obfuscator.

**Original Posted Date**: 2014-05-23T04:26:41.99-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

@jirka7a, your fix is not at the correct location. 

This bug was caused by name choosing algorithm, which mistakenly decides `IA.get_PropA` and `IB.get_PropB` should both be named to `A`. The then obfuscated assembly has a logic change that makes `ClassB.get_PropB` an overriding method of `ClassA.get_PropA`, which is of course wrong and leads to the incorrect output.

Properties can be renamed and even dropped as they are not in fact used in most of the cases. Method references usually go directly to the getter or setter methods.

A more suitable fix (instead of making `uniqueMemberNameIndex` global) is to let `IA` and `ClassA` be put into consideration when renaming `IB.get_PropB` and `ClassB.get_PropB`. Then Obfuscar can easily tell `A` is not suitable, and switch to `a` (the next unique name). 

Thus, only a few lines are needed to be inserted to `Obfuscator.GetNameGroups` method.

The fix is coming in Obfuscar 2.0.2 release.

**Original Posted Date**: 2015-04-17T20:03:53.613-07:00

**Comment by lextudio-support** [comment]
**Original Closed Date**: 2015-04-17T20:16:05.55-07:00
